handler/src/main/java/io/netty/handler/pcap/PCAPWriteHandler.java

```diff
@@ -0,0 +1,410 @@
+            int bytes = packet.readableBytes();
+
+            try {
+                if (isWriteOperation) {
+                    TCPPacket.writePacket(tcpBuf, packet, sendSegmentNumber, receiveSegmentNumber, srcAddr.getPort(),
+                            dstAddr.getPort(), TCPPacket.TCPFlag.ACK);
+                    completeTCPWrite(srcAddr, dstAddr, tcpBuf, ctx.alloc(), ctx);
+                    logTCP(true, bytes, sendSegmentNumber, receiveSegmentNumber, srcAddr, dstAddr, false);
+
+                    sendSegmentNumber += bytes;
+
+                    TCPPacket.writePacket(tcpBuf, null, receiveSegmentNumber, sendSegmentNumber, dstAddr.getPort(),
+                            srcAddr.getPort(), TCPPacket.TCPFlag.ACK);
+                    completeTCPWrite(dstAddr, srcAddr, tcpBuf, ctx.alloc(), ctx);
+                    logTCP(true, bytes, sendSegmentNumber, receiveSegmentNumber, dstAddr, srcAddr, true);
+                } else {
+                    TCPPacket.writePacket(tcpBuf, packet, receiveSegmentNumber, sendSegmentNumber, dstAddr.getPort(),
+                            srcAddr.getPort(), TCPPacket.TCPFlag.ACK);
+                    completeTCPWrite(dstAddr, srcAddr, tcpBuf, ctx.alloc(), ctx);
+                    logTCP(false, bytes, receiveSegmentNumber, sendSegmentNumber, dstAddr, srcAddr, false);
+
+                    receiveSegmentNumber += bytes;
+
+                    TCPPacket.writePacket(tcpBuf, null, sendSegmentNumber, receiveSegmentNumber, srcAddr.getPort(),
+                            dstAddr.getPort(), TCPPacket.TCPFlag.ACK);
+                    completeTCPWrite(srcAddr, dstAddr, tcpBuf, ctx.alloc(), ctx);
+                    logTCP(false, bytes, sendSegmentNumber, receiveSegmentNumber, srcAddr, dstAddr, true);
+                }
+            } finally {
+                tcpBuf.release();
+            }
+
+        } else if (msg instanceof DatagramPacket) {
+            handleUDP(ctx, ((DatagramPacket) msg).duplicate());
+        } else {
+            logger.error("Discarding Pcap Write for Object {}", msg);
+        }
+    }
+
+    private void logTCP(boolean isWriteOperation, int bytes, int sendSegmentNumber, int receiveSegmentNumber,
+                        InetSocketAddress srcAddr, InetSocketAddress dstAddr, boolean ackOnly) {
+        if (ackOnly) {
+            logger.debug("Writing TCP ACK, isWriteOperation {}, Segment Number {}, Ack Number {}, Src Addr {}, "
+                    + "Dst Addr {}", isWriteOperation, sendSegmentNumber, receiveSegmentNumber, dstAddr, srcAddr);
+        } else {
+            logger.debug("Writing TCP Data of {} Bytes, isWriteOperation {}, Segment Number {}, Ack Number {}, " +
+                            "Src Addr {}, Dst Addr {}", bytes, isWriteOperation, sendSegmentNumber,
+                    receiveSegmentNumber, srcAddr, dstAddr);
+        }
+    }
+
+    private void handleUDP(ChannelHandlerContext ctx, DatagramPacket datagramPacket) {
```

Comment:
use `handleUDP(ChannelHandlerContext ctx, ByteBuf content, InetSocketAddress src, InetSocketAddress dst)` as signature so it also can be used with "connected datagram channel" (See above)

Reply:
1 doubt, DatagramChannel will be sending DatagramPacket and each DatagramPacket has it's own Source and Destination IP. Am I right? So we don't have to depend on `ctx.channel().localAddress()` or `ctx.channel().remoteAddress()`.

Reply:
@hyperxpro yes when "not connected" that's true... but when a `DatagramChannel` is connected you can send `ByteBuf` and in this case the local / remote address is the one of the `Channel`

Reply:
Handled the signature part in a different fashion.